Ruff

ryneeverett opened the pull request:
Use ruff, per the suggestion in #1151.

ryneeverett committed:
flake8 -> ruff check

ryneeverett committed:
ruff format: test

Also, clean ups including some same-line string literal concatenations,
at least one of which was introduced by `ruff format`. I found the
others with the imperfect but good enough
`rg "[\"'].*[\"'][\s]*\+[\s]*[\"']"`.

Lotram commented on pyproject.toml:
```diff
@@ -97,6 +97,13 @@ todoist = "bugwarrior.services.todoist:TodoistService"
+
+[tool.ruff.lint]
+# Default value plus pycodestyle warnings.
+select = ["E4", "E7", "E9", "F", "W"]
```
Consider adding "I" (for isort), so imports are automatically sorted

Lotram commented on pyproject.toml:
```diff
@@ -97,6 +97,13 @@ todoist = "bugwarrior.services.todoist:TodoistService"
 bugwarrior = "bugwarrior.config.ini2toml_plugin:activate"
 
+[tool.ruff.format]
+quote-style = "single"
```
ruff default value, and the usual python standard (statement open to debate) is to use double quotes, though there is nothing wrong with using single quotes I guess

ryneeverett replied:
We're mostly using single quotes at the present. This tweak reduces the diff from >5k lines to >3k lines. But I also see the advantage of going with "the usual" if that's what people are more used to these days...

ryneeverett replied:
By switching to "preserve" the diff is reduced to >2k lines. I'm not sure there's much of an advantage to enforcing project-wide quote consistency but I'm open to revisiting the question.

ryneeverett committed:
ruff format

This commit is the sole result of running `ruff format` with
ruff-0.14.4. To verify this:

1. Check out the previous commit.
2. Run `ruff format`.
3. Run `git diff <this-commit>`. There should be no output and the
   return code should be zero.

ryneeverett commented on bugwarrior/notifications.py:
```diff
@@ -4,11 +4,9 @@
 logo_path = cache_dir + "/logo.png"
-logo_url = "https://upload.wikimedia.org/wikipedia/" + \
-    "en/5/59/Taskwarrior_logo.png"
+logo_url = "https://upload.wikimedia.org/wikipedia/" + "en/5/59/Taskwarrior_logo.png"
```
Kind of silly to leave it like this.

ryneeverett commented:
While skimming through the diff I made a number of adjustments to the configuration which further reduced the diff by aligning better with the existing style (and my own sensibilities). There's a few things I still don't care for -- newlines before closing parenthesis/brackets, indentation/newline for every nested parenthesis/bracket, spaces around operators -- but overall it seems like an improvement to the style by catching all the inconsistencies.

I ended up being very glad I chose to "preserve" the existing single/double quotes because it would have been hard to skim with all that distraction and the skimming led me not only to the configuration adjustments but to being able to see the places where cleanup from the formatting transformation (and preemptive adjustments before it) made sense. If we were to unify the quote style in the future there would be little need to review the diff.

@Lotram I'm sure you'd prefer a vanilla configuration, but I think this is a reasonable compromise for now. I'm happy to revisit changing any of these settings in the future, but I'd want feedback from other contributors. Unless you have further comments I will merge as is so i can unblock your other proposals in #1151.

Lotram commented:
> @Lotram I'm sure you'd prefer a vanilla configuration, but I think this is a reasonable compromise for now. I'm happy to revisit changing any of these settings in the future, but I'd want feedback from other contributors. Unless you have further comments I will merge as is so i can unblock your other proposals in #1151.

Indeed, even tough I would have preferred a cleaner config, that's still a huge benefit, and I understand your point. 

To be honest, I know a lot of people try to limit the number of diffs as much as possible, I don't really see the point when that's a commit we won't review anyway (because it's just formatting, and I trust the formatters to do their job without breaking the code), except maybe to avoid adding noise when using git blame / git history. 

But I don't contribute to open source that much, so I may not have all the right use-cases in mind, and you're the maintainer here, so I'm ok with this.

ryneeverett commented:
Thanks for agreeing to the compromise.

> To be honest, I know a lot of people try to limit the number of diffs as much as possible, I don't really see the point when that's a commit we won't review anyway (because it's just formatting, and I trust the formatters to do their job without breaking the code), except maybe to avoid adding noise when using git blame / git history.

I actually did review this diff, though not thoroughly. My point is that if we were to unify the quotations in the future, *that* would be a diff I would not review. While ruff didn't break the code, it did make some formatting worse like `"some string" + "another string"` on the same line or leaving continuations of dictionary items unindented which I addressed in the third commit.
